### PR TITLE
Backward function for torch.cdist

### PR DESCRIPTION
**Pull request by ifedan**
CPU and GPU implementation for backward function for torch.cdist

### CODE REVIEW

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp [hunk 1]
```diff
@@ -241,43 +241,109 @@ struct PDist {
...
+    const int64_t r1 = t1.size(-2);
+    const int64_t r2 = t2.size(-2);
+    const int64_t m = t1.size(-1);
+    const int64_t gs = grad.stride(1);
+
+    const scalar_t * const grad_start = grad.data<scalar_t>();
+    const scalar_t * const dist_start = dist.data<scalar_t>();
+    const scalar_t * const t1_start = t1.data<scalar_t>();
+    const scalar_t * const t2_start = t2.data<scalar_t>();
+    scalar_t * const res_start = result.data<scalar_t>();;
```
Comment (vadimkantorov) [inline aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: nit: double semi-colon :)

**Review by gchanan — COMMENTED** [review]
tests?

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp (hunk 1 above)
Comment (ifedan) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: Fixed

File: aten/src/ATen/native/Distance.cpp [hunk 2]
```diff
@@ -52,6 +53,21 @@ Tensor cdist(const Tensor& x1, const Tensor& x2, const double p) {
   return result;
 }
 
+Tensor _cdist_backward(const Tensor& grad, const Tensor& x1, const Tensor& x2, const double p, const Tensor& dist) {
```
[the remark this answers is not in the document]
Comment (ifedan) [reply aten/src/ATen/native/Distance.cpp]: Fixed

File: aten/src/ATen/native/Distance.cpp [hunk 3]
```diff
@@ -52,6 +53,21 @@ Tensor cdist(const Tensor& x1, const Tensor& x2, const double p) {
...
+  AT_CHECK(x1.is_contiguous(), "_cdist_backward requires X1 to be contiguous");
+  AT_CHECK(x2.is_contiguous(), "_cdist_backward requires X2 to be contiguous");
+  AT_CHECK(dist.is_contiguous(), "_cdist_backward requires dist to be contiguous");
+  int64_t n = x1.size(-2);
+  int64_t m = x1.size(-1);
+  auto device1 = x1.type().device_type();
+  AT_CHECK(device1 == kCPU || device1 == kCUDA, "_cdist_backward only supports CPU and CUDA devices, X1 got: ", device1);
+  auto device2 = x2.type().device_type();
+  AT_CHECK(device2 == kCPU || device2 == kCUDA, "_cdist_backward only supports CPU and CUDA devices, X2 got: ", device2);
+  Tensor result = at::empty({n, m}, x1.options());
```
[the remark this answers is not in the document]
Comment (ifedan) [reply aten/src/ATen/native/Distance.cpp]: Fixed

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp [hunk 4]
```diff
@@ -241,43 +241,109 @@ struct PDist {
...
+      res_vec.store(res, count);
     }
   }
 
 };
 
 void pdist_forward_kernel_impl(Tensor& result, const Tensor& self, const double p) {
   AT_DISPATCH_FLOATING_TYPES(self.type(), "pdist", [&] {
-    PDist<scalar_t>::apply(result, self, p);
+    PDist<scalar_t>::apply_pdist(result, self, p);
```
[the remark this answers is not in the document]
Comment (ifedan) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp]: Fixed

**Comment by ifedan** [comment]
> tests?

Added tests

**Comment by vadimkantorov** [comment]
@t-vi noted a slow pdist backward in https://lernapparat.de/jit-optimization-intro, not sure if it was reported anywhere

**Pull request closed without merging**